File: js/src/vista-sdk/lib/internal/LocalIdParsingErrorBuilder.ts
import { ParsingState } from "../types/LocalId";
import { Error, ErrorBuilder } from "./ErrorBuilder";

export class LocalIdParsingErrorBuilder extends ErrorBuilder<ParsingState> {
    readonly predefinedMessages: Map<ParsingState, string>;

    public static readonly Empty : LocalIdParsingErrorBuilder = new LocalIdParsingErrorBuilder();

    public static Create() : LocalIdParsingErrorBuilder{
        return new LocalIdParsingErrorBuilder();
    }
    constructor(errors?: Error<ParsingState>[]) {
        super(errors);
        this.predefinedMessages = this.setPredefinedMessages();
    }

    public get hasError(): boolean {
        return this.errors.length > 0;
    }

    private setPredefinedMessages(): Map<ParsingState, string> {
        const map = new Map<ParsingState, string>();

        map.set(ParsingState.NamingRule, "Missing or invalid naming rule");
        map.set(ParsingState.VisVersion, "Missing or invalid vis version");
        map.set(ParsingState.PrimaryItem, "Missing or invalid primary item");
        map.set(ParsingState.SecondaryItem, "Invalid secondary item");
        map.set(ParsingState.ItemDescription, "Missing or invalid /meta prefix");
        map.set(ParsingState.MetaQuantity, "Invalid metadata tag: Quantity");
        map.set(ParsingState.MetaContent, "Invalid metadata tag: Content");
        map.set(ParsingState.MetaCommand, "Invalid metadata tag: Command");
        map.set(ParsingState.MetaPosition, "Invalid metadata tag: Position");
        map.set(ParsingState.MetaCalculation, "Invalid metadata tag: Calculation");
        map.set(ParsingState.MetaState, "Invalid metadata tag: State");
        map.set(ParsingState.MetaType, "Invalid metadata tag: Type");
        map.set(ParsingState.MetaDetail, "Invalid metadata tag: Detail");
        map.set(ParsingState.EmptyState, "Missing primary path or metadata");

        return map;
    }
}

File: csharp/src/Vista.SDK/Internal/LocalIdParsingErrorBuilder.cs
namespace Vista.SDK.Internal;

internal sealed record LocalIdParsingErrorBuilder
{
    private readonly List<(LocalIdParsingState type, string message)> _errors;
    private static Dictionary<LocalIdParsingState, string> _predefinedErrorMessages =>
        SetPredefinedMessages();

    internal static readonly LocalIdParsingErrorBuilder Empty = new();

    internal LocalIdParsingErrorBuilder() => _errors = new List<(LocalIdParsingState, string)>();

    internal LocalIdParsingErrorBuilder AddError(LocalIdParsingState state)
    {
        if (!_predefinedErrorMessages.TryGetValue(state, out var predefinedMessage))
            throw new Exception("Couldn't find predefined message for: " + state.ToString());

        _errors.Add((state, predefinedMessage));
        return this;
    }

    internal LocalIdParsingErrorBuilder AddError(LocalIdParsingState state, string? message)
    {
        if (string.IsNullOrWhiteSpace(message))
            return AddError(state);

        _errors.Add((state, message!));
        return this;
    }

    internal bool HasError => _errors.Count > 0;

    internal static LocalIdParsingErrorBuilder Create() => new();

    public ParsingErrors Build() =>
        _errors.Count == 0
            ? ParsingErrors.Empty
            : new ParsingErrors(_errors.Select((t, m) => (t.type.ToString(), t.message)).ToArray());

    private static Dictionary<LocalIdParsingState, string> SetPredefinedMessages()
    {
        var parsingMap = new Dictionary<LocalIdParsingState, string>
        {
            { LocalIdParsingState.NamingRule, "Missing or invalid naming rule" },
            { LocalIdParsingState.VisVersion, "Missing or invalid vis version" },
            {
                LocalIdParsingState.PrimaryItem,
                "Invalid or missing Primary item. Local IDs require atleast primary item and 1 metadata tag."
            },
            { LocalIdParsingState.SecondaryItem, "Invalid secondary item" },
            { LocalIdParsingState.ItemDescription, "Missing or invalid /meta prefix" },
            { LocalIdParsingState.MetaQuantity, "Invalid metadata tag: Quantity" },
            { LocalIdParsingState.MetaContent, "Invalid metadata tag: Content" },
            { LocalIdParsingState.MetaCommand, "Invalid metadata tag: Command" },
            { LocalIdParsingState.MetaPosition, "Invalid metadata tag: Position" },
            { LocalIdParsingState.MetaCalculation, "Invalid metadata tag: Calculation" },
            { LocalIdParsingState.MetaState, "Invalid metadata tag: State" },
            { LocalIdParsingState.MetaType, "Invalid metadata tag: Type" },
            { LocalIdParsingState.MetaDetail, "Invalid metadata tag: Detail" },
            { LocalIdParsingState.EmptyState, "Missing primary path or metadata" }
        };
        return parsingMap;
    }
}
